File: odk/examples/DevelopersGuide/Charts/CalcHelper.java
/*************************************************************************
 *
 *  The Contents of this file are made available subject to the terms of
 *  the BSD license.
 *
 *  Copyright 2000, 2010 Oracle and/or its affiliates.
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *  1. Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *  2. Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *  3. Neither the name of Sun Microsystems, Inc. nor the names of its
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
 *  OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 *  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 *  TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 *  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *************************************************************************/

// __________ Imports __________

import java.util.Random;

import com.sun.star.uno.UnoRuntime;
import com.sun.star.lang.XMultiServiceFactory;
// container access
import com.sun.star.container.*;

// application specific classes
import com.sun.star.sheet.*;
import com.sun.star.table.*;
import com.sun.star.chart.*;
import com.sun.star.text.XText;

import com.sun.star.document.XEmbeddedObjectSupplier;
import com.sun.star.frame.XModel;
// base graphics things
import com.sun.star.awt.Point;
import com.sun.star.awt.Size;
import com.sun.star.awt.Rectangle;

// Exceptions
import com.sun.star.uno.RuntimeException;
import com.sun.star.container.NoSuchElementException;

// __________ Implementation __________

/** Helper for accessing a calc document
    @author Bj&ouml;rn Milcke
 */
public class CalcHelper
{
    public CalcHelper( XSpreadsheetDocument aDoc )
    {
        maSpreadSheetDoc = aDoc;
        initSpreadSheet();
    }

    // ____________________

    public XSpreadsheet getChartSheet() throws RuntimeException
    {
        XNameAccess aSheetsNA = (XNameAccess) UnoRuntime.queryInterface(
            XNameAccess.class, maSpreadSheetDoc.getSheets() );

        XSpreadsheet aSheet = null;
        try
        {
            aSheet = (XSpreadsheet) UnoRuntime.queryInterface(
                XSpreadsheet.class, aSheetsNA.getByName( msChartSheetName ) );
        }
        catch( NoSuchElementException ex )
        {
            System.out.println( "Couldn't find sheet with name " + msChartSheetName + ": " + ex );
        }
        catch( Exception ex )
        {}

        return aSheet;
    }

    // ____________________

    public XSpreadsheet getDataSheet() throws RuntimeException
    {
        XNameAccess aSheetsNA = (XNameAccess) UnoRuntime.queryInterface(
            XNameAccess.class, maSpreadSheetDoc.getSheets() );

        XSpreadsheet aSheet = null;
        if( aSheetsNA != null )
        {
            try
            {
                aSheet = (XSpreadsheet) UnoRuntime.queryInterface(
                    XSpreadsheet.class, aSheetsNA.getByName( msDataSheetName ) );
            }
            catch( NoSuchElementException ex )
            {
                System.out.println( "Couldn't find sheet with name " + msDataSheetName + ": " + ex );
            }
            catch( Exception ex )
            {}
        }

        return aSheet;
    }

    // ____________________

    /** Insert a chart using the given name as name of the OLE object and the range as correspoding
        range of data to be used for rendering.  The chart is placed in the sheet for charts at
        position aUpperLeft extending as large as given in aExtent.

        The service name must be the name of a diagram service that can be instantiated via the
        factory of the chart document
     */
    public XChartDocument insertChart(
        String               sChartName,
        CellRangeAddress     aRange,
        Point                aUpperLeft,
        Size                 aExtent,
        String               sChartServiceName )
    {
        XChartDocument aResult = null;
        XTableChartsSupplier aSheet;

        // get the sheet to insert the chart
        try
        {
            aSheet = (XTableChartsSupplier) UnoRuntime.queryInterface(
                XTableChartsSupplier.class, getChartSheet() );
        }
        catch( Exception ex )
        {
            System.out.println( "Sheet not found" + ex );
            return aResult;
        }

        XTableCharts aChartCollection = aSheet.getCharts();
        XNameAccess  aChartCollectionNA = (XNameAccess) UnoRuntime.queryInterface(
            XNameAccess.class, aChartCollection );

        if( aChartCollectionNA != null &&
            ! aChartCollectionNA.hasByName( sChartName ) )
        {
            Rectangle aRect = new Rectangle( aUpperLeft.X, aUpperLeft.Y, aExtent.Width, aExtent.Height );

            CellRangeAddress[] aAddresses = new CellRangeAddress[ 1 ];
            aAddresses[ 0 ] = aRange;

            // first bool: ColumnHeaders
            // second bool: RowHeaders
            aChartCollection.addNewByName( sChartName, aRect, aAddresses, true, false );

            try
            {
                XTableChart aTableChart = (XTableChart) UnoRuntime.queryInterface(
                    XTableChart.class, aChartCollectionNA.getByName( sChartName ));

                // the table chart is an embedded object which contains the chart document
                aResult = (XChartDocument) UnoRuntime.queryInterface(
                    XChartDocument.class,
                    ((XEmbeddedObjectSupplier) UnoRuntime.queryInterface(
                        XEmbeddedObjectSupplier.class,
                        aTableChart )).getEmbeddedObject());

                // create a diagram via the factory and set this as new diagram
                aResult.setDiagram(
                    (XDiagram) UnoRuntime.queryInterface(
                        XDiagram.class,
                        ((XMultiServiceFactory) UnoRuntime.queryInterface(
                            XMultiServiceFactory.class,
                            aResult )).createInstance( sChartServiceName )));
            }
            catch( NoSuchElementException ex )
            {
                System.out.println( "Couldn't find chart with name " + sChartName + ": " + ex );
            }
            catch( Exception ex )
            {}
        }

        return aResult;
    }

    // ____________________

    /** Fill a rectangular range with random numbers.
        The first column has increasing values
     */
    public XCellRange insertRandomRange( int nColumnCount, int nRowCount )
    {
        XCellRange aRange = null;

        // get the sheet to insert the chart
        try
        {
            XSpreadsheet aSheet = getDataSheet();
            XCellRange aSheetRange = (XCellRange) UnoRuntime.queryInterface( XCellRange.class, aSheet );

            aRange = aSheetRange.getCellRangeByPosition(
                0, 0,
                nColumnCount - 1, nRowCount - 1 );

            int nCol, nRow;
            double fBase  = 0.0;
            double fRange = 10.0;
            double fValue;
            Random aGenerator = new Random();


            for( nCol = 0; nCol < nColumnCount; nCol++ )
            {
                if( 0 == nCol )
                {
                    (aSheet.getCellByPosition( nCol, 0 )).setFormula( "X" );
                }
                else
                {
                    (aSheet.getCellByPosition( nCol, 0 )).setFormula( "Random " + nCol );
                }

                for( nRow = 1; nRow < nRowCount; nRow++ )
                {
                    if( 0 == nCol )
                    {
                        // x values: ascending numbers
                        fValue = (double)nRow + aGenerator.nextDouble();
                    }
                    else
                    {
                        fValue = fBase + ( aGenerator.nextGaussian() * fRange );
                    }

                    // put value into cell

                    // note: getCellByPosition is a method at ...table.XCellRange which
                    //       the XSpreadsheet inherits via ...sheet.XSheetCellRange
                    (aSheet.getCellByPosition( nCol, nRow )).setValue( fValue );
                }
            }

        }
        catch( Exception ex )
        {
            System.out.println( "Sheet not found" + ex );
        }

        return aRange;
    }

    // ____________________

    public XCellRange insertFormulaRange( int nColumnCount, int nRowCount )
    {
        XCellRange aRange = null;

        // get the sheet to insert the chart
        try
        {
            XSpreadsheet aSheet = getDataSheet();
            XCellRange aSheetRange = (XCellRange) UnoRuntime.queryInterface( XCellRange.class, aSheet );

            aRange = aSheetRange.getCellRangeByPosition(
                0, 0,
                nColumnCount - 1, nRowCount - 1 );

            int nCol, nRow;
            double fValue;
            double fFactor = 2.0 * java.lang.Math.PI / (double)(nRowCount - 1);
            String aFormula;

            // set variable factor for cos formula
            int nFactorCol = nColumnCount + 2;
            (aSheet.getCellByPosition( nFactorCol - 1, 0 )).setValue( 0.2 );

            XText xCellText = (XText) UnoRuntime.queryInterface( XText.class, aSheet.getCellByPosition( nFactorCol - 1, 1 ) );
            xCellText.setString( "Change the factor above and\nwatch the changes in the chart" );

            for( nCol = 0; nCol < nColumnCount; nCol++ )
            {
                for( nRow = 0; nRow < nRowCount; nRow++ )
                {
                    if( 0 == nCol )
                    {
                        // x values: ascending numbers
                        fValue = (double)nRow * fFactor;
                        (aSheet.getCellByPosition( nCol, nRow )).setValue( fValue );
                    }
                    else
                    {
                        aFormula = new String( "=" );
                        if( nCol % 2 == 0 )
                            aFormula += "SIN";
                        else
                            aFormula += "COS";
                        aFormula += "(INDIRECT(ADDRESS(" + (nRow + 1) + ";1)))+RAND()*INDIRECT(ADDRESS(1;" + nFactorCol + "))";
                        (aSheet.getCellByPosition( nCol, nRow )).setFormula( aFormula );
                    }
                }
            }

        }
        catch( Exception ex )
        {
            System.out.println( "Sheet not found" + ex );
        }

        return aRange;
    }

    // ____________________

    /** Bring the sheet containing charts visually to the foreground
     */
    public void raiseChartSheet()
    {
        ((XSpreadsheetView) UnoRuntime.queryInterface(
            XSpreadsheetView.class,
            ((XModel) UnoRuntime.queryInterface(
                XModel.class,
                maSpreadSheetDoc )).getCurrentController()) ).setActiveSheet( getChartSheet() );
    }


    // __________ private members and methods __________

    private final String  msDataSheetName  = "Data";
    private final String  msChartSheetName = "Chart";

    private XSpreadsheetDocument   maSpreadSheetDoc;


    // ____________________

    /** create two sheets, one for data and one for charts in the document
     */
    private void initSpreadSheet()
    {
        if( maSpreadSheetDoc != null )
        {
            XSpreadsheets  aSheets    = maSpreadSheetDoc.getSheets();
            XNameContainer aSheetsNC  = (XNameContainer)  UnoRuntime.queryInterface(
                XNameContainer.class, aSheets );
            XIndexAccess   aSheetsIA  = (XIndexAccess)    UnoRuntime.queryInterface(
                XIndexAccess.class, aSheets );

            if( aSheets   != null &&
                aSheetsNC != null &&
                aSheetsIA != null )
            {
                try
                {
                    // remove all sheets except one
                    for( int i = aSheetsIA.getCount() - 1; i > 0; i-- )
                    {
                        aSheetsNC.removeByName(
                            ( (XNamed) UnoRuntime.queryInterface(
                                XNamed.class, aSheetsIA.getByIndex( i ) )).getName() );
                    }

                    XNamed aFirstSheet = (XNamed) UnoRuntime.queryInterface(
                        XNamed.class,
                        aSheetsIA.getByIndex( 0 ));

                    // first sheet becomes data sheet
                    aFirstSheet.setName( msDataSheetName );

                    // second sheet becomes chart sheet
                    aSheets.insertNewByName( msChartSheetName, (short)1 );
                }
                catch( Exception ex )
                {
                    System.out.println( "Couldn't initialize Spreadsheet Document: " + ex );
                }
            }
        }
    }
}

File: odk/examples/DevelopersGuide/Charts/python/CalcHelper.py
# -*- tab-width: 4; indent-tabs-mode: nil; py-indent-offset: 4 -*-
#
# This file is part of the LibreOffice project.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#

import math
import random
import sys
import traceback

from com.sun.star.awt import Rectangle
from com.sun.star.container import NoSuchElementException


class CalcHelper:

    _data_sheet_name = "Data"
    _chart_sheet_name = "Chart"

    def __init__(self, doc):
        self.spread_sheet_doc = doc
        self._init_spread_sheet()

    def get_chart_sheet(self):
        sheets = self.spread_sheet_doc.Sheets
        try:
            sheet = sheets.getByName(self._chart_sheet_name)
        except NoSuchElementException as err:
            print(f"Couldn't find sheet with name {self._chart_sheet_name}: {err}")
            traceback.print_exc()
        except Exception as err:
            print(err)
            traceback.print_exc()
        return sheet

    def get_data_sheet(self):
        sheets = self.spread_sheet_doc.Sheets
        try:
            sheet = sheets.getByName(self._data_sheet_name)
        except NoSuchElementException as err:
            print(f"Couldn't find sheet with name {self._data_sheet_name}: {err}")
            traceback.print_exc()
        except Exception as err:
            print(err)
            traceback.print_exc()
        return sheet

    def insert_chart(self, chart_name, range_, upper_left, extent, chart_service_name):
        """Insert a chart using the given name as name of the OLE object and the range as corresponding
           range of data to be used for rendering.  The chart is placed in the sheet for charts at
           position upper_left extending as large as given in extent.

           The service name must be the name of a diagram service that can be instantiated via the
           factory of the chart document

        Args:
            chart_name (str): _description_
            range_ (com.sun.star.table.CellRangeAddress): _description_
            upper_left (Point): _description_
            extent (Size): _description_
            chart_service_name (str): _description_
        """
        result = None
        try:
            sheet = self.get_chart_sheet()
        except Exception as err:
            print(f"Sheet not found {err}")
            traceback.print_exc()
            return

        chart_collection = sheet.getCharts()

        if not chart_collection.hasByName(chart_name):
            rect = Rectangle(upper_left.X, upper_left.Y, extent.Width, extent.Height)
            addresses = []
            addresses.append(range_)

            # first bool: ColumnHeaders
            # second bool: RowHeaders
            chart_collection.addNewByName(chart_name, rect, addresses, True, False)

            try:
                table_chart = chart_collection.getByName(chart_name)

                # the table chart is an embedded object which contains the chart document
                result = table_chart.getEmbeddedObject()

                # create a diagram via the factory and set this as new diagram
                result.setDiagram(result.createInstance(chart_service_name))

            except NoSuchElementException as err:
                print(f"Couldn't find chart with name {chart_name}: {err}")
                traceback.print_exc()
                return

            except Exception as err:
                print(err)
                traceback.print_exc()
                return

        return result

    def insert_random_range(self, column_count, row_count):
        """Fill a rectangular range with random numbers.
           The first column has increasing values

        Args:
            column_count (int): _description_
            row_count (int): _description_

        Return:
            (com.sun.star.table.XCellRange)
        """
        cell_range = None
        try:
            # get the sheet to insert the chart
            sheet = self.get_data_sheet()
            cell_range = sheet[0:row_count, 0:column_count]

            base = 0.0
            float_range = 10.0

            for col in range(column_count):
                if col == 0:
                    sheet[0, col].Formula = "X"
                else:
                    sheet[0, col].Formula = f"Random {col}"

                for row in range(1, row_count):
                    if col == 0:
                        value = row + random.random()
                    else:
                        value = base + random.gauss(0.0, 1.0) * float_range

                    # put value into cell
                    sheet[row, col].Value = value

        except Exception as err:
            print(f"Sheet not found {err}")
            traceback.print_exc()

        return cell_range

    def insert_formula_range(self, column_count, row_count):
        try:
            # get the sheet to insert the chart
            sheet = self.get_data_sheet()
            cell_range = sheet[0 : row_count, 0 : column_count]
            factor = 2.0 * math.pi / (row_count - 1)

            factor_col = column_count + 2
            sheet[0, factor_col - 1].Value = 0.2
            sheet[1, factor_col - 1].String = "Change the factor above and\nwatch the changes in the chart"

            for col in range(column_count):
                for row in range(row_count):
                    if col == 0:
                        # x values: ascending numbers
                        value = row * factor
                        sheet[row, col].Value = value
                    else:
                        formula = "="
                        if col % 2 == 0:
                            formula += "SIN"
                        else:
                            formula += "COS"

                        formula += f"(INDIRECT(ADDRESS({row + 1};1)))+RAND()*INDIRECT(ADDRESS(1;{factor_col}))"
                        sheet[row, col].Formula = formula

        except Exception as err:
            print(f"Sheet not found {err}")
            traceback.print_exc()

        return cell_range

    def raise_chart_sheet(self):
        """Bring the sheet containing charts visually to the foreground"""
        self.spread_sheet_doc.getCurrentController().setActiveSheet(self.get_chart_sheet())

    def _init_spread_sheet(self):
        """create two sheets, one for data and one for charts in the document"""
        if self.spread_sheet_doc is not None:
            sheets = self.spread_sheet_doc.Sheets
            if sheets:
                for i in range(len(sheets) - 1, 0, -1):
                    sheets.removeByName(sheets.getByIndex(i).getName())
                try:
                    first_sheet = sheets[0]
                    first_sheet.setName(self._data_sheet_name)
                    sheets.insertNewByName(self._chart_sheet_name, 1)
                except Exception as e:
                    print(f"Couldn't initialize Spreadsheet Document: {e}", file=sys.stderr)
                    traceback.print_exc()

# vim: set shiftwidth=4 softtabstop=4 expandtab:
